docs/enhancements/using-dtk-for-building-driver-containers.md

```diff
@@ -0,0 +1,127 @@
+      dockerImageReference: registry.ci.openshift.org/ocp/4.11-2022-07-27-013908@sha256:484c9fe9900020e9b8f5601cbe86d054fe47a6f16e8d9dc6ecbfd58bffd1ad4b
+      generation: 4
+      image: sha256:484c9fe9900020e9b8f5601cbe86d054fe47a6f16e8d9dc6ecbfd58bffd1ad4b
+    tag: 411.86.202207260413-0
+  - items:
+    - created: "2022-07-27T02:55:02Z"
+      dockerImageReference: registry.ci.openshift.org/ocp/4.11-2022-07-27-013908@sha256:484c9fe9900020e9b8f5601cbe86d054fe47a6f16e8d9dc6ecbfd58bffd1ad4b
+      generation: 4
+      image: sha256:484c9fe9900020e9b8f5601cbe86d054fe47a6f16e8d9dc6ecbfd58bffd1ad4b
+    - created: "2022-07-27T02:20:47Z"
+      dockerImageReference: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:84edddb4322cfe50d8545c41014a8f5595a4260a20d9a269e144909c5949582b
+      generation: 2
+      image: sha256:84edddb4322cfe50d8545c41014a8f5595a4260a20d9a269e144909c5949582b
+    tag: latest
+```
+
+We can see that:
+* We have a different tag for each RHCOS version that the cluster has known.
+* The ‘latest’ tag point to the most recent RHCOS tag in the imagestream
+* The `status` field of the imagestream keeps an array of images for each tag with the full digest history used for that tag
+    * The latest tag was first pointing to `410.84.202207140725-0` tag and then move to point to `411.86.202207260413-0` tag.
+
+
+## Node reconciler flow
+Create and update an in-memory mapping `kernelVersion` → `rhcosVersion`.
+* The `kernelVersion` can be found on the node object at `node.status.nodeInfo.kernelVersion`
+* The `rhcosVersion` can be found on the node object at `node.status.nodeInfo.osImage`
+
+## Module reconciler flow
+
+When a build pod is created, or when a cluster upgrade kicks, KMMO will get the list of kernels running in the cluster from the node labels.
+For each kernel in the cluster, if a build is required, get the matching `rhcosVersion` from the `kernelVersion` → `rhcosVersion` mapping. If there is more than one we can just pick the latest RHCOS available version with the required kernel.
+KMMO will use that RHCOS version as the imagestream tag in order to build/rebuild the driver container.
+
+## Dockerfile examples
+If the user want us to pull the correct DTK image on builds/cluster-upgrade, he will need to put a placeholder in the Dockerfile for KMMO to substitute. In case this placeholder isn't present in the Dockerfile, KMMO will pull the base image specified by the user.
+
+#### Dockerfile which let KMMO find the correct DTK image.
+```
+FROM DTK-AUTO
+RUN gcc ...
+...
+```
+
+#### Dockerfile with 0-magic.
+```
+FROM user.registry.com/org/user-base-image:tag
+RUN gcc ...
+...
+```
+
+## Disconnected clusters
+This approach is expected to work also in disconnected clusters as we don’t need to pull the driver-toolkit image manually to get the kernel RPMs installed in it. Only the buildconfig will pull the DTK image, therefore, it will be redirected if the user has set an `ImageSourceContentPolicy` in the cluster.
```

Comment:
I think KMMO will anyway need to support disconnected environments like SRO does.
Let's take into consideration a vendor who ships a `Module` CR that references driver images hosted on their registry accessible on the internet. It is a lot more convenient for customers to apply that CR without any modification and tune the `ImageSourceContentPolicy` to mirror that registry vs altering URLs in the `Module`.

Reply:
In addition we will need that code for preflight also

Reply:
> I think KMMO will anyway need to support disconnected environments like SRO does.

Can you elaborate on what part will need to be implemented to support disconnected (besides reading the user CAs)?

> Let's take into consideration a vendor who ships a Module CR that references driver images hosted on their registry accessible on the internet. It is a lot more convenient for customers to apply that CR without any modification and tune the ImageSourceContentPolicy to mirror that registry vs altering URLs in the Module.

Absolutely! this is not what that section was describing - it was explaining that unlike the current implementation, we don't have to pull the relase-info image nor the DTK image in KMMO, therefore, we don't have to read `containers.conf` etc. Instead we assume that the DTK image is the correct one (thanks to the mapping) and just pass it to the buildconfig object as the base image. The buildconfig object in his turn will pull DTK, but the buildconfig should be aware of ISCP in the cluster so that remove the need of having that logic embedded into KMMO (as we do today in SRO).

Reply:
we will still need that code for preflight verification, since in preflight we do need to access the images from the code

Reply:
There is also the option of generating a job and check its output instead of pulling it directly from our code. But this is not related to that proposal.

Reply:
> Can you elaborate on what part will need to be implemented to support disconnected (besides reading the user CAs)?

We also need:
- authentication
- mirroring config (either from the `ImageSourceContentPolicy` or `registries.conf` on the node)

> Does that make better sense?

It does make sense. I was just pointing out that not using the disconnected code in the proposal is barely a benefit, because we will anyway leverage it in other parts of the operator.

Reply:
I have updated the "Disconnected" section.